var/spack/repos/builtin/packages/timemory/package.py

```diff
@@ -48,77 +92,83 @@ class Timemory(CMakePackage):
     depends_on('cuda', when='+cuda')
     depends_on('cuda', when='+cupti')
     depends_on('upcxx', when='+upcxx')
-    depends_on('gotcha', when='+gotcha')
     depends_on('likwid', when='+likwid')
-    depends_on('caliper', when='+caliper')
+    depends_on('gotcha', when='~build_gotcha+gotcha')
+    depends_on('caliper', when='~build_caliper+caliper')
+    depends_on('dyninst', when='+dyninst')
     depends_on('gperftools', when='+gperftools')
+    depends_on('intel-parallel-studio', when='+vtune')
 
     conflicts('+cupti', when='~cuda', msg='CUPTI requires CUDA')
+    conflicts('+kokkos_tools', when='~tools',
+              msg='+kokkos_tools requires +tools')
+    conflicts('+kokkos_modules', when='~tools',
+              msg='+kokkos_modules requires +tools')
+    conflicts('tls_model=local-dynamic', when='+python',
+              msg='+python require tls_model=global-dynamic')
+    conflicts('tls_model=initial-exec', when='+python',
+              msg='+python require tls_model=global-dynamic')
+    conflicts('tls_model=local-exec', when='+python',
+              msg='+python require tls_model=global-dynamic')
+    conflicts('+mpip_library', when='~mpi', msg='+mpip_library requires +mpi')
+    conflicts('+mpip_library', when='~gotcha', msg='+mpip_library requires +gotcha')
+    conflicts('+ompt_library', when='~ompt', msg='+ompt_library requires +ompt')
 
     def cmake_args(self):
         spec = self.spec
 
-        # Use spack install of Caliper and/or GOTCHA
-        # instead of internal submodule build
         args = [
-            '-DTIMEMORY_BUILD_GOTCHA=OFF',
-            '-DTIMEMORY_BUILD_CALIPER=OFF',
-            '-DTIMEMORY_BUILD_TOOLS=ON',
+            '-DTIMEMORY_BUILD_PYTHON=ON',
```

Comment:
Shouldn't this remain conditional based on whether or not `+python` is requested? Python is only a dependency when `+python` is requested.

Reply:
For any of the options w/ both USE and BUILD variants (python, caliper, gotcha, and ompt), if `TIMEMORY_USE_<PACKAGE>=OFF` then `TIMEMORY_BUILD_<PACKAGE>` is ignored. USE w/o BUILD means find external installation, USE + BUILD means build it from submodule.

Reply:
With Spack, we try to never build vendored copies of software, or to link to system libraries. If a variant is on, it should link to a Spack installation. If it's off, it shouldn't link to anything at all. If the dependency is mandatory, there shouldn't even be a variant for it. This prevents packages from linking to system libraries, leading to non-deterministic builds. It also prevents packages from being built with multiple versions of the same library in the DAG. So if I understand you correctly, we never want to build from a submodule, we either want to link to an external Spack installation or not link at all.

Reply:
So I am not trying to link to any system libraries here. "Python" in this context is actually just a template library used to generate the bindings, pybind11, and I will never let spack provide that because there is a 0% chance of it breaking any other spack build/install. 

The gotcha submodule can easily be removed, it was added bc I added the others. 

The ompt submodule can also be removed. I didn't realize there was a spack package for that which worked (`llvm-openmp-ompt`) because `ompt-openmp` is outdated. Is there a syntax for "use this config if not explicitly specified but allow another config?" E.g. if they just run `spack install timemory +ompt` then I want `depends_on("llvm-openmp-ompt+standalone")` but if they explicitly want to use `llvm-openmp-ompt~standalone` (the default), then that is fine. 

Basically the situation above is what prompted this whole submodule additions. I found it quite difficult to just do `spack install timemory +caliper` because timemory doesn't require a specific papi version but Caliper does so I would just like to defer to caliper setting the papi version. That was just the first issue, many more happened and the specs got extremely detailed, especially once TAU got involved. Basically, timemory doesn't really care about the spec of Caliper, TAU, or any similar libraries for which it provides wrappers (arm-forge, etc.). Those libraries provide generic interfaces to call their library and their corresponding timemory components just invoke those generic interfaces, e.g. `tau_marker::start()` just calls `Tau_start(...)`. Is there a way to say "if there was a previous spack install for package X, just use that installation" _without having to write a packages.yaml_? And, as mentioned previously, "use this spec as the default unless otherwise specified".

Reply:
> Is there a way to say "if there was a previous spack install for package X, just use that installation" without having to write a packages.yaml?

This will be the default behavior of the new concretizer. An opt-in experimental version of this concretizer is expected to be released sometime in May (Spack 0.15) with a full replacement of the old concretizer scheduled for Spack 0.16.

> And, as mentioned previously, "use this spec as the default unless otherwise specified".

Not currently, the `depends_on` directive is a hard requirement, there is no way to specify preferences. We've talked about adding this, but it's harder than it sounds. For now, I would let all `depends_on` statements be as relaxed as possible.

Reply:
Thanks. I believe all the above is resolved now